# Tarea 1

_Tarea 1_ de _Benjamín Rivera_ para el curso de __Métodos Numéricos__ impartido por _Joaquín Peña Acevedo_. Fecha limite de entrega __6 de Septiembre de 2020__.

## Como ejecutar


#### Requerimientos

Este programa se ejecuto en mi computadora con la version de __Python 3.8.2__ y con estos
[requerimientos](https://github.com/BenchHPZ/UG-Compu/blob/master/MN/requerimientos.txt)

### Jupyter

En caso de tener acceso a un _servidor jupyter_ ,con los requerimientos antes mencionados, unicamente basta con ejecutar todas las celdas de este _notebook_. Probablemente no todas las celdas de _markdown_ produzcan el mismo resultado por las 
[_Nbextensions_](jupyter-contrib-nbextensions.readthedocs.io).

### Consola

Habrá archivos e instrucciones para poder ejecutar cada uno de los ejercicios desde la consola.

### Si todo sale mal

En caso de que todo salga mal, tratare de dejar una copia disponible en __GoogleColab__ que se pueda ejecutar con la versión de __Python__ de _GoogleColab_

## Ejercicio 1

Supongamos que una computadora tiene 8 dígitos para representar la parte fraccionaria
de un número de punto flotante.

### Calcular el valor del épsilon de la máquina

En clase vimos que si conocemos la cantidad de bits $p$ para representar la parte fraccionaria de la mantisa, entonces 

\begin{equation} \epsilon_m = (1.0)_2 \times 2^{-p} \label{eq:epsilon}\end{equation}

Dado que sabemos que la máquina de este ejercicio usara $8bits$ para representar la parte fraccionaria entonces, por \ref{eq:epsilon}, el $\epsilon_m$ para este ejercicio es

\begin{equation*} \epsilon_m = (1.0)_2 \times 2^{-8} = 3.90625 \times 10^{-3}
\label{eq:epsilon maquina} \end{equation*}

In [1]:
print(2**(-8))

0.00390625


### Dar la representación en notación cientifica (mantisa base 2, multiplicada por 2 elevda al exponente correspondiente) del número 5.

Sabemos que el número $5$ se representa en binario como $(101)_2$. Como los números se prefieren normalizados entonces debemos representar $(1.01)_2$ en la notación solicitada. Por lo que 

\begin{equation*} 5 = (101)_2 \times 2^0 = (1.01)_2 \times 2^2 = 1.25 \times 2^2
\end{equation*}

### Dar la representación científica del número consecutico a 5 en la computadora. Escribir la distancia $d_c$ entre 5 y su consecutivo. Expresar $d_c$ en términos del épslon de la máquina. 

Sabemos que el número consecutivo $(5_c)$ a $5$ en esta computadora es $5+\epsilon_m$. Si extendemos toda la mantisa de $5$ este se ve $(1.01000000)$. Y sumar $\epsilon_m$ implica sumar $1$ a la mantisa, enspecificamente en esta computadora, al $bit 8$ de la mantisa. De manera que 

\begin{eqnarray*} 
    5_c &=& 5 + \epsilon_m 
    = (1.01)_2 \times 2^2 + (1.0)_2 \times 2^{2-8} \\ 
    &=& ((1.01000000)_2 + (0.00000001)_2) \times 2^2 \\
    &=& (1.01000001)_2 \times 2^2 \\
    &=& 5.015625
\end{eqnarray*}

por lo que la distancia entre $5$ y su consecutivo $5+\epsilon_m$ es

\begin{equation*} d_c = (5+\epsilon_m)-5 = \epsilon_m \times 2^2 = (0.00000001)_2 \times 2^2 = 0.015625
\end{equation*}

### Tenemos que el consecutivo de $5$ es expresable como $5 + d_c$. Si tenemos un $x$ real tal que $x \in (5, 5+d_c)$ entonces la computadora representara a $x$ como $fl(x)=5$ o $fl(x)=5+d_c$. Escribir una cota para el error relativo para las dos posibles represetnaciones de x

> Error relativo
\begin{equation*}
    \left| \frac{fl(x) - x}{x} \right|
\end{equation*}

Sabemos que las dos posibles representaciones son el _truncamiento_ y el _redondeo_ ,para dar la cota se deben calcular los valores minimos y maximos del error relativo. En general trabajaremos con

\begin{equation} \label{eq:error relativo}
    \left| \frac{fl(x) - x}{x} \right| = \left| \frac{fl(x)}{x} - 1 \right| 
\end{equation}

Antes de continuar es importante notar lo siguiente. Sabemos que, para ambos casos, el dominio de la función sera $[5, 5+d_c]$. Además, notemos que el rango de $\frac{fl(x)}{x}$, para nuestro dominio y los dos valores que puede tomar $fl(x)$, es $\left[ \frac{5}{5+d_c}, \frac{5+d_c}{5} \right]$. Por ultimo, sabemos que $\frac{fl(x)}{x}$ es descendente para $x>0$.

__Truncamiento__ Para el truncamiento $fl(x) = 5$ por lo que debemos calcular la cota para 

\begin{equation*} Er_\_ = \left| \frac{5}{x} - 1 \right|
\end{equation*}

En este caso, el rango de $\frac{fl(x)}{x} = \frac{5}{x}$ es $\left[ \frac{5}{5+d_c}, \frac{5}{5}=1 \right]$ por lo que el rango de $Er_\_$ es $\left[ 0, |\frac{5}{5+d_c}-1| \right]$. Por iluminación divina, suponemos que esta función es ascendente, y por lo tanto $Er_\_(5) < Er_\_(5+dc)$. Primero evaluamos la función en estos puntos

\begin{equation*} Er_\_(5) = \left|\frac{5}{5} -1\right| = 0 \quad\quad Er_\_(5+d_c) = \left|\frac{5}{5+d_c} -1 \right| \sim 0.00311
\end{equation*}

Por lo que, usando el truncamiento, esta __<ins>función esta acotada</ins>__ por $\left[ Er_\_(5), Er_\_(5+d_c) \right]$


__Redondeo__ Procedemos de manera similar al anterior. Para el redondeo tenemos que $fl(x) = 5+d_c$, esto nos da la expresion

\begin{equation*} Er_+ = \left| \frac{5+d_c}{x} - 1 \right|
\end{equation*}

Y de manera similar al anterior podemos ver que esta función es descendente, por lo que $Er_+(5+d_c) < Er_+(d_c)$

\begin{equation*} Er_+(5+d_c) = \left| \frac{5+d_c}{5+d_c} -1 \right| = 0 \quad\quad Er_+(5) = \left| \frac{5+d_c}{5} -1 \right| \sim 0.00312
\end{equation*}

De manera que esta función queda acotada por $\left[ Er_+(5+d_c),Er_+(5) \right]$

### Explique si los siguientes números tienen respresentación exacta en la computadora, es decir, si $fl(a_i)=a_i$

 - $a_1 = \epsilon/2$
 \par Sabemos que en el sistema de este ejercicio ($8bits$ parte flotante) $\epsilon_m = (1.0)_2 \times 2^{-8}$, de manera que,
 
 \begin{equation*} \epsilon_m/2 = \frac{(1.0)_2 \times 2^{-8}}{2} = (1.0)_2 \times 2^{-9}
 \end{equation*}
 
 por lo que, mientras que el rango del exponente sea suficiente (dado que el ejercicio solo da información del tamaño de la \textit{matisa}), este número __es representable__ en el sistema.


 - $a_2 = 1+ \epsilon/2$ 
 \par De manera que $a_2$ se expresa como
 
 \begin{equation*} a_2 = (1.0)_2 + \epsilon/2 = (1.0)_2 + (1.0)_2 \times 2^{-9} = (1.000000001)_2
 \end{equation*}
 
 pero como este sistema solo usa $8bits$ para la mantisa entonces, para este sistema
 
 \begin{equation*} a_2 = (1.0)_2 + \epsilon/2 = (1.000000001)_2 = (1.0)_2
 \end{equation*}
 
 por lo que este número __no tiene representación__ en este sistema.


 - $a_3 = 1- \epsilon$
 \par Este número se escribe como
 
 \begin{eqnarray*} 
     a_3 = 1- \epsilon &=& (1.0)_2 - (1.0)_2 \times 2^{-8} \\
         &=& (1.00000000)_2 - (0.00000001)_2 \\
         &=& (0.11111111)_2 \\
         &=& (1.1111111)_2 \times 2^{-1}
 \end{eqnarray*}
 
 por lo que este número __si tiene una representación__ en este sistema.


 - $a_4 = 1- \epsilon/2$
 \par Como con los anteriores, esta operación se expresa como
 
 \begin{eqnarray*}
     a_4 = 1- \epsilon/2 &=& (1.0)_2 - (1.0)_2 \times 2^{-9} \\
         &=& (1.00000000)_2 - (0.000000001)_2 \\
         &=& (0.111111111)_2 \\
         &=& (1.11111111)_2 \times 2^{-1}
 \end{eqnarray*}
 
 el cual __si tiene respresentación__ en el sistema de este ejercicio.
 
 
 - $a_5 = 1- \epsilon/4$
 \par En este inciso primero calcularemos $\epsilon/4$, para el cual expandimos lo siguiente
 
 \begin{equation*} \epsilon/4 = \frac{(1.0)_2 \times 2^{-8}}{2^2} = (1.0)_2 \times 2^{-10}
 \end{equation*}
 
 el cual, mientras el exponente alcance, _si es representable_ en el sistema. Por otro lado, el numero de este inciso nos da
 
 \begin{eqnarray*}
     a_5 = 1 - \epsilon/4 &=& (1.0)_2 - (1.0)_2 \times 2^{-10} \\
         &=& (1.00000000)_2 - (0.0000000001)_2 \\
         &=& (0.1111111111)_2 \\
         &=& (1.111111111)_2 \times 2^{-1}
 \end{eqnarray*}
 
 pero este número tiene una mantisa de $9bits$, por lo cual, __no tiene representación__ en el sistema.


 - $a_6 = \epsilon^2$
 \par Para este inciso
 
 \begin{equation} a_6 = \epsilon^2 = ((1.0)_2 \times 2^{-8})_2^2 = (1.0)_2 \times 2^{-16} 
 \end{equation}
 
 el cual __si tiene represetación__ en el sistema.
 
 
 - $a_7 = 0.125$
 \par Primero pasamos de el número de decimal a binario, de manera que $(0.125)_{10} = (0.001)_{2}$. Luego hay que normalizarlo, por lo que $(0.001)_{2} = (1.0)_{2} \times 2^{-3}$. Por lo que este numero __si tiene respresentacion__ en el sistema.
 
 
 - $a_8 = 2^{-10}$
 \par Y por \'ultimo, tenemos que
 
 \begin{equation*} a_8 = 2^{-10} = ((10.0)_2)^{-10} = ((1.0)_2 \times 2^{1})^{-10} = (1.0)_2 \times 2^{-10}
 \end{equation*}


### De una cota para el error relativo de las restas \ref{eq:resta 1} y \ref{eq:resta 2} respecto al verdader valor. Suponga que $fl(x)$ se obtiene por redondeo hacia abajo _(truncamiento)_

\begin{equation}
    fl(0.9) - fl(0.5)   \label{eq:resta 1}
\end{equation} \begin{equation}
    fl(0.9) - fl(0.895) \label{eq:resta 2}
\end{equation}

Para este ejercicio, dado que nos pide usar el \textit{truncamiento}, definimos a la \textbf{unidad de redondeo} como $u = \epsilon_m/2$. Esto tambien nos define a $fl(x)=x(1+\delta)$ y $|\delta| \leq u$. Adem\'as, siguiendo las notas, podemos usar la cota que ah\'i se proporciona y sustituir $u$.

\begin{equation} |\delta_{x-y}| \leq \frac{\epsilon}{2}\frac{|x|+|y|}{|x-y|} \label{eq:cota}
\end{equation}

Y antes de continuar calculamos $\epsilon_m/2$ para poder usarlo mas adelante. De manera que 

\begin{equation*} \epsilon_m/2 = \frac{(1.0)_2 \times 2^{-8}}{2} = (1.0)_2 \times 2^{-9}
\end{equation*}

Y, como tambien lo usaremos, es bueno tener en cuenta.

\begin{equation*} \delta_{x-y} = \frac{fl(x)-fl(y) - (x-y)}{x-y}
\end{equation*}

 Empezamos por la operaci\'on \ref{eq:resta 1}, de donde obtenemos $x=0.9$ y $y=0.5$. Primero calculamos los valores de maquina de estos.
 \begin{eqnarray*}
     fl(x) = fl(0.9) &=& (1.1100(1100))_2\times 2^{-1} \\
             &=& (1.11001100)_2\times 2^{-1}  \quad\quad\quad \text{(Truncamiento (no cabe))} \\
             &\sim& 0.8984375
            &&\\&&\\
     fl(y) = fl(0.5) &=& (1.0)_2\times 2^{-1} \\
             &=& (1.00000000)_2\times 2^{-1}  \quad\quad\quad \text{(Truncamiento (si cabe))} \\
             &=& 0.5
 \end{eqnarray*}
 
 De manera que, como ya conocemos $\epsilon_m$, los valores de la operaci\'on y sus representaciones en el sistema, procedemos a encontrar la cota. Dado que usamos la cota de las notas, unicamente queda sustituir, de esto obtenemos que: 
 
\begin{eqnarray*}
    \delta_{x-y} \sim -0.00390625 \\
    \frac{\epsilon}{2}\frac{|x|+|y|}{|x-y|} = 0.00683593
\end{eqnarray*}

Por lo que, siguiendo la cota de la ecuaci\'on \ref{eq:cota}, podemos decir que esta operaci\'on esta acotada por

\begin{equation*}
    |\delta_{x-y}| \sim 0.00390625 \leq 0.00683593 \sim \frac{\epsilon}{2}\frac{|x|+|y|}{|x-y|}
\end{equation*}

De manera similar, para la operaci\'on \ref{eq:resta 2}, tenemos que $x=0.9$ y $y=0.895$. Calculamos los redondondeos de la computadora.

 \begin{eqnarray*}
     fl(x) = fl(0.9) &=& (1.1100(1100))_2\times 2^{-1} \\
             &=& (1.11001100)_2\times 2^{-1}  \quad\quad\quad \text{(Truncamiento (no cabe))} \\
             &\sim& 0.8984375
            &&\\&&\\
     fl(y) = fl(0.895) &=& (1.1100101000111101)_2\times 2^{-1} \\
             &=& (1.11001010)_2\times 2^{-1}  \quad\quad\quad \text{(Truncamiento (no cabe))} \\
             &=& 0.89453125
 \end{eqnarray*}
 
Ahora procedemos a calcular los limites de la cota, por lo que

\begin{eqnarray*}
    \delta_{x-y} \sim  -0.21875\\
    \frac{\epsilon}{2}\frac{|x|+|y|}{|x-y|} = 0.70117187
\end{eqnarray*}

Y por \'ultimo, seg\'un  la cota \ref{eq:cota}, acotamos esta operaci\'on por

\begin{equation*}
    |\delta_{x-y}| \sim -0.21875 \leq 0.70117187 \sim \frac{\epsilon}{2}\frac{|x|+|y|}{|x-y|}
\end{equation*}

In [2]:
# Funciones auxiliares para el Ejercicio 2.3
def deltaXY(fx, fy, x, y):
    return ((fx-fy)-(x-y))/(x-y)

def cotaSup(x, y):
    ep2 = 0.001953125
    return ep2*(abs(x) + abs(y))/abs(x - y)

#print(deltaXY(0.8984375, 0.89453125, 0.9, 0.895))
#print(cotaSup(0.9, 0.895))

## Ejercicio 2

Programe la función `epsilonFloat` que devuelve el épsilon de la máquina $\epsilon_m$ para números de simple precisión y la función `epsilonDouble` para números de doble precisión.
```
epsilon = 1.0
unidad = 1.0
valor = unidad + epsilon
while valor > unidad
    epsilon = epsilon/2
    valor = end + epsilon
end while
epsilon = epsilon*2
```
Usar el algoritmo visto en clase.

In [3]:
# Primera parte del ejercicio 1

import numpy as np

def epsilonMaquina(tipoDato):
    """ Funcion que trata de calcular el epsilon 
    de la maquina mediante el algoritmo antes pre_
    sentado. 
    Input:
        tipoDato := esta pensado para ser uno de 
            los tipos proporcionados por la li_
            breria numpy.
    Output:
        Regresa el epsilon de la maquina calcula_
            do con el tipo de dato especificado.
    """
    epsilon = tipoDato(1.0)
    unidad = tipoDato(1.0)
    valor = unidad + epsilon
    
    while valor > unidad:
        epsilon = epsilon/tipoDato(2.0)
        valor = unidad + epsilon
        
    return epsilon*2


def epsilonFloat():
    """ Calculamos el epsilon de la maquina con 
    precision de 32bits
    """
    return epsilonMaquina(np.float32)

def epsilonDouble():
    """ Calculamos el epsilon de la maquina con
    precision de 64 bits
        A pesar de que el flotante de python ya
    tiene esta precision, creo que es convenien_
    te especificarlo.
    """
    return epsilonMaquina(np.float64)

# Epsilons calculados
eF = np.float32(epsilonFloat())
eD = np.float64(epsilonDouble())

# Imprimir en pantalla
print(f'Se calculalron los epsilons \n\t{eF=} y \n\t{eD=} \n'
      'para 32 y 64 bits correspondientemente.\n')

Se calculalron los epsilons 
	eF=1.1920929e-07 y 
	eD=2.220446049250313e-16 
para 32 y 64 bits correspondientemente.



In [4]:
# Segunda parte del ejercicio 1

def respuesta(res):
    """ Funcion para formatear la respuesta. """
    return "iguales" if res else "diferentes"

def comparacion(epsilon):
    """ Esta funcion resivira el epsilon a eva_
    luar y el tipo de dato al que este correspon_
    de para hacer las comparaciones solicitadas
    en el ejercicio.
    Input:
        En caso de que 
    Output:
        Las respuestas son procesadas por la
        funcion respuesta para obtener el for_
        mato solicitado
            True implica que son iguales
            False implica que son diferentes
    """
    tD = type(epsilon) # Para escribir menos
    print(f'Con {epsilon=} y tipo de dato = {str(tD)} se da que')
    
    # Comprobaciones
    print(f'{respuesta( tD(1 + epsilon )   == 1 )  =}')
    print(f'{respuesta( tD( epsilon/2 )    == 0 )  =}')
    print(f'{respuesta( tD(1 + epsilon/2 ) == 1 )  =}')
    print(f'{respuesta( tD(1 - epsilon/2 ) == 1 )  =}')
    print(f'{respuesta( tD(1 - epsilon/4 ) == 1 )  =}')
    print(f'{respuesta( tD( epsilon**2 )   == 0 )  =}')
    print(f'{respuesta(epsilon + tD(epsilon**2) == epsilon) =}')
    print(f'{respuesta(epsilon - tD(epsilon**2) == epsilon) =}')
    
    print('...\n')
    
# Hacemos la comparacion para 32bits
comparacion(eF)
# y para 64
comparacion(eD)

Con epsilon=1.1920929e-07 y tipo de dato = <class 'numpy.float32'> se da que
respuesta( tD(1 + epsilon )   == 1 )  ='diferentes'
respuesta( tD( epsilon/2 )    == 0 )  ='diferentes'
respuesta( tD(1 + epsilon/2 ) == 1 )  ='iguales'
respuesta( tD(1 - epsilon/2 ) == 1 )  ='diferentes'
respuesta( tD(1 - epsilon/4 ) == 1 )  ='iguales'
respuesta( tD( epsilon**2 )   == 0 )  ='diferentes'
respuesta(epsilon + tD(epsilon**2) == epsilon) ='diferentes'
respuesta(epsilon - tD(epsilon**2) == epsilon) ='diferentes'
...

Con epsilon=2.220446049250313e-16 y tipo de dato = <class 'numpy.float64'> se da que
respuesta( tD(1 + epsilon )   == 1 )  ='diferentes'
respuesta( tD( epsilon/2 )    == 0 )  ='diferentes'
respuesta( tD(1 + epsilon/2 ) == 1 )  ='iguales'
respuesta( tD(1 - epsilon/2 ) == 1 )  ='diferentes'
respuesta( tD(1 - epsilon/4 ) == 1 )  ='iguales'
respuesta( tD( epsilon**2 )   == 0 )  ='diferentes'
respuesta(epsilon + tD(epsilon**2) == epsilon) ='diferentes'
respuesta(epsilon - tD(epsilon**2) == 

De manera que el programa calculo que el _epsilon de la maquina_ es {{eF}} y {{eD}} para las precisiones de _32_ y _64 bits_ correspondientemente, y repecto a las comparaciones unicamente se encontraron __dos igualdades__ cuando se uso preciosion de _64bits_.

#### Como ejecutar

[GoogleColab](https://colab.research.google.com/gist/BenchHPZ/cd5bc176f3e3841fa1e84924feca9ec2/tarea1-benjamin_rivera-metodos_numericos.ipynb)
<a href="https://colab.research.google.com/gist/BenchHPZ/cd5bc176f3e3841fa1e84924feca9ec2/tarea1-benjamin_rivera-metodos_numericos.ipynb">
    <img src="../../../assets/colab-badge.svg" 
         alt="Open In Colab"/>
</a>

Para ejecutar este ejercicio en __consola__ es importante ubicarse en la misma carpeta del 
[archivo `T1.py`](https://github.com/BenchHPZ/UG-Compu/blob/master/MN/Tareas/T1/T1.py)
y ejecutar el siguiente comando en consola
```console
python3 T1.py
```
Este programa no espera recibir argumento alguno. La salida debe ser similar a la siguiente imagen


![Evidencia Ejecucion](assets/E2-1.png)